src/core/lib/event_engine/cf_engine/dns_service_resolver.cc

```diff
@@ -0,0 +1,222 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <grpc/support/port_platform.h>
+
+#ifdef GPR_APPLE
+
+#include "absl/strings/str_format.h"
+
+#include "src/core/lib/address_utils/parse_address.h"
+#include "src/core/lib/event_engine/cf_engine/dns_service_resolver.h"
+#include "src/core/lib/event_engine/posix_engine/lockfree_event.h"
+#include "src/core/lib/event_engine/tcp_socket_utils.h"
+#include "src/core/lib/event_engine/trace.h"
+#include "src/core/lib/gprpp/host_port.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+void DNSServiceResolverImpl::LookupHostname(
+    EventEngine::DNSResolver::LookupHostnameCallback on_resolve,
+    absl::string_view name, absl::string_view default_port) {
+  GRPC_EVENT_ENGINE_DNS_TRACE(
+      "DNSServiceResolverImpl::LookupHostname: name: %.*s, default_port: %.*s, "
+      "this: %p",
+      static_cast<int>(name.length()), name.data(),
+      static_cast<int>(default_port.length()), default_port.data(), this);
+
+  absl::string_view host;
+  absl::string_view port_string;
+  if (!grpc_core::SplitHostPort(name, &host, &port_string)) {
+    engine_->Run([on_resolve = std::move(on_resolve),
+                  status = absl::InvalidArgumentError(
+                      absl::StrCat("Unparseable name: ", name))]() mutable {
+      on_resolve(status);
+    });
+    return;
+  }
+  GPR_ASSERT(!host.empty());
+  if (port_string.empty()) {
+    if (default_port.empty()) {
+      engine_->Run([on_resolve = std::move(on_resolve),
+                    status = absl::InvalidArgumentError(absl::StrFormat(
+                        "No port in name %s or default_port argument",
+                        name))]() mutable { on_resolve(std::move(status)); });
+      return;
+    }
+    port_string = default_port;
+  }
+
+  int port = 0;
+  if (port_string == "http") {
+    port = 80;
+  } else if (port_string == "https") {
+    port = 443;
+  } else if (!absl::SimpleAtoi(port_string, &port)) {
+    engine_->Run([on_resolve = std::move(on_resolve),
+                  status = absl::InvalidArgumentError(absl::StrCat(
+                      "Failed to parse port in name: ", name))]() mutable {
+      on_resolve(std::move(status));
+    });
+    return;
+  }
+
+  // TODO(yijiem): Change this when refactoring code in
+  // src/core/lib/address_utils to use EventEngine::ResolvedAddress.
+  grpc_resolved_address addr;
+  const std::string hostport = grpc_core::JoinHostPort(host, port);
+  if (grpc_parse_ipv4_hostport(hostport.c_str(), &addr,
+                               /*log_errors=*/false) ||
+      grpc_parse_ipv6_hostport(hostport.c_str(), &addr,
+                               /*log_errors=*/false)) {
+    // Early out if the target is an ipv4 or ipv6 literal.
```

Comment:
aha ok thanks for explaining!

In that case, if there's a way for us to detect whether the passed param is a "ip literal" v.s. not, do we still need the former method?

I'm not suggesting that we need to do that, but just curious about whether it would be possible to reduce dependency on 2 sets of implementations, and what is the cost of that (since logic to detect whether something is a "ip literal" can be as simple as a regex match :))

Reply:
That's c-core util functions, I expect depending on them should be fine, no?
It's more than regex as we are also validating an ip address, need to consider the invalid values, broad cast addresses, etc. IPv6 can be even more complex.

Reply:
It is fine to depend on it! I was just trying to understand the reasons and necessity for it. If it saves meaningful complexity to reuse their code, then it's totally valid to do so!

Maybe you could add a comment explaining what case are you relying on them v.s. the iOS solutions? So at least we have awareness on how to simplify if there's ever a need in the future. 😃